File: engine/app/schemas/trade.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class TradeResponse(BaseModel):
    id: int
    import_batch_id: int | None
    source: str
    ticket: str
    symbol: str
    direction: str
    open_time: datetime
    close_time: datetime
    open_price: float
    close_price: float
    volume: float
    profit: float
    commission: float
    swap: float
    net_profit: float
    stop_loss: float | None
    take_profit: float | None
    close_reason: str | None

    model_config = {"from_attributes": True}


class TradeDeleteRequest(BaseModel):
    trade_ids: list[int] = Field(min_length=1)

    @model_validator(mode="after")
    def validate_trade_ids(self) -> "TradeDeleteRequest":
        if len(set(self.trade_ids)) != len(self.trade_ids):
            raise ValueError("交易不能重复。")
        return self


class TradeDeleteResponse(BaseModel):
    deleted_count: int


class TradeDateRangeResponse(BaseModel):
    from_time: datetime | None = None
    to_time: datetime | None = None


class TradeListResponse(BaseModel):
    items: list[TradeResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/trade.ts
export type TradeSource = "MT5" | "CSV";

export interface Trade {
  id: number;
  import_batch_id: number | null;
  source: TradeSource;
  ticket: string;
  symbol: string;
  direction: string;
  open_time: string;
  close_time: string;
  open_price: number;
  close_price: number;
  volume: number;
  profit: number;
  commission: number;
  swap: number;
  net_profit: number;
  stop_loss: number | null;
  take_profit: number | null;
  close_reason: string | null;
}

export interface TradeDateRange {
  from_time: string | null;
  to_time: string | null;
}

export interface TradeListResponse {
  items: Trade[];
  total: number;
  page: number;
  page_size: number;
}
